Milestone 5: Friendships v1

WinterCore opened the pull request:
Changelog:

- Create friendships database table.
- Create friendship model.
- Add models associations between `User` and `Friendship`.
- Add unit tests for models' associations and validations.
- Add integration tests for listing, creating, accepting, and rejecting friendships.

Bismarck-GM committed:
Add friendship associations. Rspec first test

Bismarck-GM committed:
Add tests for models

WinterCore committed:
Add friendships send/accept/reject controller methods

Bismarck-GM committed:
Add test for friendship model

Bismarck-GM committed:
Add tests for changes in models

Bismarck-GM committed:
Fix rubocop offenses

WinterCore committed:
Merge branch 'development' into feature/friendshipV1

WinterCore committed:
Add friendships integration tests

Bismarck-GM committed:
Fix bug

Bismarck-GM committed:
Merge branch 'feature/friendshipV1' of github.com:Bismarck-GM/ror-social-scaffold into feature/friendshipV1

Bismarck-GM committed:
Add link helper to show accept/cancel friend requests

kobiyoyo commented on app/models/user.rb:
```diff
@@ -9,4 +9,24 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships
+  has_many :reverse_friendships, foreign_key: :friend_id, class_name: Friendship.name
```
This can be written like this `has_many :reverse_friendships, foreign_key: :friend_id, class_name: "Friendship"`

kobiyoyo reviewed (changes requested):
## Hi Team  ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
* [x] Linters config: Do not make any changes in config files, when done lets fix any errors that might show up.
* [x] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] Lets write an integration test for the following components:
- User
- Comment
- Post 
- Like
## [OPTIONAL]
- [x] Make sure there are no `n + 1` queries in index views [Recommended article](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5).  
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

WinterCore committed:
Add changes requested by TSE

- Filter timeline posts to posts created by the logged in user and their
  friends
- Add integration & unit tests for models, friendships, and the
  timeline,
- Add reject friendships
- Add comments & likes counters

WinterCore committed:
Fix rubocop offences

WinterCore committed:
Fix bug with user timeline (Timeline was showing posts of unaccepted friend requests)

Dandush03 commented:
# Status: :rocket: Project Approved :rocket: 

# Great work implementing changes, keep up the good work

# Happy coding :nerd_face: :computer:
